### PR TITLE
MCLOUD-5953: Backport Redis patches

### PR DESCRIPTION
**Pull request by BaDos**
### Fixed Issues (if relevant)
https://jira.corp.magento.com/browse/MCLOUD-5953

### Manual testing scenarios

Need to test for Magento 2.3.3 and 2.3.4
1. Edit `composer.json`
```
"repositories": {
        "repo2": {
            "type": "composer",
            "url": "https://repo.magento.com/"
        },
        "ece-tools": {
            "type": "vcs",
            "url": "https://github.com/magento/ece-tools.git"
        },
        "mcp": {
            "type": "vcs",
            "url": "git@github.com:magento/magento-cloud-patches.git"
        },
        "docker":{
            "type": "vcs",
            "url": "git@github.com:magento/magento-cloud-docker.git"
        }
  },
```
```
"require": {
        "magento/magento-cloud-metapackage": ">=2.3.3 <2.3.4",
        "magento/ece-tools": "dev-develop as 2002.1.99",
        "magento/magento-cloud-patches": "dev-MCLOUD-5953 as 1.0.99"
},
```

2. Run `composer update`
3. Commit changes of `composer.json` and `composer.lock`
4. Push to trigger redeploy
5. Check that frontend works
6. Check `app/etc/env.php`
```
'cache' =>
  array (
    'frontend' =>
    array (
      'default' =>
      array (
        'backend' => 'Cm_Cache_Backend_Redis',
        'backend_options' =>
        array (
          'server' => 'redis.internal',
          'port' => 6379,
          'database' => 1,
        ),
        'id_prefix' => 'f53_',
      ),
      'page_cache' =>
      array (
        'backend' => 'Cm_Cache_Backend_Redis',
        'backend_options' =>
        array (
          'server' => 'redis.internal',
          'port' => 6379,
          'database' => 2,
        ),
        'id_prefix' => 'f53_',
      ),
    ),
  ),
```
7. Add `.magento.env.yaml`
```
stage:
  deploy:
    CACHE_CONFIGURATION:
      _merge: false
      frontend:
        default:
          backend: '\Magento\Framework\Cache\Backend\RemoteSynchronizedCache'
          backend_options:
            remote_backend: '\Magento\Framework\Cache\Backend\Redis'
            remote_backend_options:
              persistent: 0
              server: 'redis.internal'
              database: '1'
              port: '6379'
              password: ''
              compress_data: '1'
            local_backend: 'Cm_Cache_Backend_File'
            local_backend_options:
              cache_dir: '/dev/shm/'
          frontend_options:
            write_control: false
      type:
        default:
          frontend: 'default'
```
8. Commit  `.magento.env.yaml`
9. Push to trigger redeploy
10. Check that frontend works
11. Check `app/etc/env.php`
```
'cache' =>
  array (
    'frontend' =>
    array (
      'default' =>
      array (
        'backend' => '\\Magento\\Framework\\Cache\\Backend\\RemoteSynchronizedCache',
        'backend_options' =>
        array (
          'remote_backend' => '\\Magento\\Framework\\Cache\\Backend\\Redis',
          'remote_backend_options' =>
          array (
            'persistent' => 0,
            'server' => 'redis.internal',
            'database' => '1',
            'port' => '6379',
            'password' => '',
            'compress_data' => '1',
          ),
          'local_backend' => 'Cm_Cache_Backend_File',
          'local_backend_options' =>
          array (
            'cache_dir' => '/dev/shm/',
          ),
        ),
        'frontend_options' =>
        array (
          'write_control' => false,
        ),
      ),
    ),
    'type' =>
    array (
      'default' =>
      array (
        'frontend' => 'default',
      ),
    ),
  ),
```

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: patches.json [hunk 1]
```diff
@@ -241,8 +241,14 @@
...
       ">=2.3.2 <2.3.5": "MCLOUD-3025__sitemap_generation_warnings__2.3.3.patch"
+    },
+    "Cache Locking performance issue": {
+      ">=2.3.4 <2.3.5": "MDVA-26795__fix_performance_issue_in_cache_locking_mechanism__2.3.4.patch"
+    },
+    "Large amount of Keys Block Cache": {
+      ">=2.3.1 <2.3.5": "MDVA-22950__large_amount_of_keys_block_cache__2.3.1.patch"
     }
-  },
+    },
```
Comment (BarnyShergold) [inline patches.json]: Remove space

File: patches/MDVA-22950__large_amount_of_keys_block_cache__2.3.1.patch [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+-use Magento\Catalog\Model\Layer\Category as CategoryLayer;
+ use Magento\ConfigurableProduct\Helper\Data;
+ use Magento\ConfigurableProduct\Model\ConfigurableAttributeData;
+ use Magento\Customer\Helper\Session\CurrentCustomer;
+@@ -154,7 +153,7 @@ class Configurable extends \Magento\Swatches\Block\Product\Renderer\Configurable
+         $this->unsetData('allow_products');
+         return parent::getJsonConfig();
+     }
+-
++
```
Comment (BarnyShergold) [inline patches/MDVA-22950__large_amount_of_keys_block_cache__2.3.1.patch]: Remove extra line break

File: patches/MDVA-26795__fix_performance_issue_in_cache_locking_mechanism__2.3.4.patch [hunk 3]
```diff
@@ -0,0 +1,835 @@
...
++++ b/vendor/magento/framework/App/Cache/Frontend/Pool.php
+@@ -152,6 +152,15 @@ class Pool implements \Iterator
+         if (isset($this->_instances[$identifier])) {
+             return $this->_instances[$identifier];
+         }
+-        throw new \InvalidArgumentException("Cache frontend '{$identifier}' is not recognized.");
++
++        if (!isset($this->_instances[self::DEFAULT_FRONTEND_ID])) {
++            throw new \InvalidArgumentException(
++                "Cache frontend '{$identifier}' is not recognized. As well as " .
```
Comment (BarnyShergold) [inline patches/MDVA-26795__fix_performance_issue_in_cache_locking_mechanism__2.3.4.patch]: Don't embed variables in strings - use sprintf if required

File: patches/MDVA-26795__fix_performance_issue_in_cache_locking_mechanism__2.3.4.patch [hunk 4]
```diff
@@ -0,0 +1,835 @@
...
++    }
++
++    /**
++     * Load value with given id from cache
++     *
++     * @param  string  $id                     Cache id
++     * @param  boolean $doNotTestCacheValidity If set to true, the cache validity won't be tested
++     * @return bool|string
++     */
++    public function load($id, $doNotTestCacheValidity = false)
```
Comment (BarnyShergold) [inline patches/MDVA-26795__fix_performance_issue_in_cache_locking_mechanism__2.3.4.patch:275]: Return type after prototype missing

File: patches/MDVA-26795__fix_performance_issue_in_cache_locking_mechanism__2.3.4.patch [hunk 5]
```diff
@@ -0,0 +1,835 @@
...
++    public function __construct($options = [])
++    {
++        $this->preloadKeys = $options['preload_keys'] ?? [];
++        parent::__construct($options);
++    }
++
++    /**
++     * Load value with given id from cache
++     *
++     * @param  string  $id                     Cache id
```
Comment (BarnyShergold) [inline patches/MDVA-26795__fix_performance_issue_in_cache_locking_mechanism__2.3.4.patch]: Remove extra whitespace

File: patches/MDVA-26795__fix_performance_issue_in_cache_locking_mechanism__2.3.4.patch [hunk 6]
```diff
@@ -0,0 +1,835 @@
...
++    /**
++     * Cover errors on save operations, which may occurs when Redis cannot evict keys, which is expected in some cases.
++     *
++     * @param string $data
++     * @param string $id
++     * @param array $tags
++     * @param bool $specificLifetime
++     * @return bool
++     */
++    public function save($data, $id, $tags = [], $specificLifetime = false)
```
Comment (BarnyShergold) [inline patches/MDVA-26795__fix_performance_issue_in_cache_locking_mechanism__2.3.4.patch:304]: Return type after prototype missing

File: patches/MDVA-26795__fix_performance_issue_in_cache_locking_mechanism__2.3.4.patch [hunk 7]
```diff
@@ -0,0 +1,835 @@
...
+diff -Nuar a/vendor/magento/framework/Cache/Backend/RemoteSynchronizedCache.php b/vendor/magento/framework/Cache/Backend/RemoteSynchronizedCache.php
+--- a/vendor/magento/framework/Cache/Backend/RemoteSynchronizedCache.php
++++ b/vendor/magento/framework/Cache/Backend/RemoteSynchronizedCache.php
+@@ -9,8 +9,10 @@ namespace Magento\Framework\Cache\Backend;
+ /**
+  * Remote synchronized cache
+  *
+- * This class created for correct work local caches with multiple web nodes,
+- * that will be check cache status from remote cache
++ * This class created for correct work witch local caches and multiple web nodes,
```
Comment (BarnyShergold) [inline patches/MDVA-26795__fix_performance_issue_in_cache_locking_mechanism__2.3.4.patch:324]: witch SHOULD BE with?

File: patches/MDVA-26795__fix_performance_issue_in_cache_locking_mechanism__2.3.4.patch [hunk 8]
```diff
@@ -0,0 +1,835 @@
...
++ * in order to be sure that we always have up to date local version of cache.
++ * This class will be check cache version from remote cache and in case it newer
```
Comment (BarnyShergold) [inline patches/MDVA-26795__fix_performance_issue_in_cache_locking_mechanism__2.3.4.patch:326]: insert 'is' after it

File: patches/MDVA-26795__fix_performance_issue_in_cache_locking_mechanism__2.3.4.patch [hunk 9]
```diff
@@ -0,0 +1,835 @@
...
++ * than local one, it will update local one from remote cache a.k.a two level cache.
```
Comment (BarnyShergold) [inline patches/MDVA-26795__fix_performance_issue_in_cache_locking_mechanism__2.3.4.patch:327]: Please don't use acronyms in comments - some non-native English speakers might not know what it means

File: patches/MDVA-26795__fix_performance_issue_in_cache_locking_mechanism__2.3.4.patch [hunk 10]
```diff
@@ -0,0 +1,835 @@
...
++
++    /**
++     * Return hash sign of the data.
+      *
+-     * @return void
++     * @param string $data
++     * @return string
+      */
+-    private function updateRemoteCacheStatusInfo()
++    private function getDataVersion(string $data)
```
Comment (BarnyShergold) [inline patches/MDVA-26795__fix_performance_issue_in_cache_locking_mechanism__2.3.4.patch:377]: Return type after prototype missing

File: patches/MDVA-26795__fix_performance_issue_in_cache_locking_mechanism__2.3.4.patch [hunk 11]
```diff
@@ -0,0 +1,835 @@
...
+
+     /**
+-     * {@inheritdoc}
++     * Load data version by id from remote.
++     *
++     * @param string $id
++     * @return false|string
+      */
+-    public function setDirectives($directives)
++    private function loadRemoteDataVersion(string $id)
```
Comment (BarnyShergold) [inline patches/MDVA-26795__fix_performance_issue_in_cache_locking_mechanism__2.3.4.patch:392]: Return type after prototype missing

File: patches/MDVA-26795__fix_performance_issue_in_cache_locking_mechanism__2.3.4.patch [hunk 12]
```diff
@@ -0,0 +1,835 @@
...
+-     * {@inheritdoc}
++     * Save new data version to remote.
++     *
++     * @param string $data
++     * @param string $id
++     * @param array $tags
++     * @param mixed $specificLifetime
++     * @return bool
++     */
++    private function saveRemoteDataVersion(string $data, string $id, array $tags, $specificLifetime = false)
```
Comment (BarnyShergold) [inline patches/MDVA-26795__fix_performance_issue_in_cache_locking_mechanism__2.3.4.patch:410]: Return type after prototype missing

File: patches/MDVA-26795__fix_performance_issue_in_cache_locking_mechanism__2.3.4.patch [hunk 13]
```diff
@@ -0,0 +1,835 @@
...
++        return $this->remote->save($this->getDataVersion($data), $id . self::HASH_SUFFIX, $tags, $specificLifetime);
++    }
++
++    /**
++     * Remove remote data version.
++     *
++     * @param string $id
++     * @return bool
++     */
++    private function removeRemoteDataVersion($id)
```
Comment (BarnyShergold) [inline patches/MDVA-26795__fix_performance_issue_in_cache_locking_mechanism__2.3.4.patch]: Return type after prototype missing

File: patches/MDVA-26795__fix_performance_issue_in_cache_locking_mechanism__2.3.4.patch [hunk 14]
```diff
@@ -0,0 +1,835 @@
...
++        ConfigWriterInterface $configWriter
++    ) {
++        $this->moduleReader = $moduleReader;
++        $this->configWriter = $configWriter;
++    }
++
++    /**
++     * @inheritDoc
++     */
++    public function doOperation()
```
Comment (BarnyShergold) [inline patches/MDVA-26795__fix_performance_issue_in_cache_locking_mechanism__2.3.4.patch]: Return type after prototype missing

File: patches/MDVA-26795__fix_performance_issue_in_cache_locking_mechanism__2.3.4.patch [hunk 15]
```diff
@@ -0,0 +1,835 @@
...
++        $this->configWriter->write(
++            'app_action_list',
++            $actionList
++        );
++    }
++
++    /**
++     * @inheritDoc
++     */
++    public function getName()
```
Comment (BarnyShergold) [inline patches/MDVA-26795__fix_performance_issue_in_cache_locking_mechanism__2.3.4.patch]: Return type after prototype missing

File: patches/MDVA-26795__fix_performance_issue_in_cache_locking_mechanism__2.3.4.patch [hunk 16]
```diff
@@ -0,0 +1,835 @@
...
++    {
++        return 'App action list generation';
```
Comment (BarnyShergold) [inline patches/MDVA-26795__fix_performance_issue_in_cache_locking_mechanism__2.3.4.patch:797]: Should be a constant

File: patches/MDVA-26795__fix_performance_issue_in_cache_locking_mechanism__2.3.4.patch (hunk 15 above)
Comment (BarnyShergold) [inline patches/MDVA-26795__fix_performance_issue_in_cache_locking_mechanism__2.3.4.patch]: Where does this method ever get called?

**Review by BarnyShergold — CHANGES_REQUESTED** [review]
Please add description from ticket MCLOUD-5953 to this PR's description. Not all Reviewers have access to the JIRA ticket.

**Review by oshmyheliuk — APPROVED** [review]

**Review by arhiopterecs — APPROVED** [review]

**Comment by andriyShevtsov** [comment]
QA approved

**Comment by andriyShevtsov** [comment]
@BarnyShergold Thank you for review. Unfortunately we are not able to change patches. It was provided by core team and already a part of magento next release. Changes in patches on our side may broke something.

**Comment by shiftedreality** [comment]
@BarnyShergold, please approve

**Review by BarnyShergold — APPROVED** [review]
Accepted with acknowledgement to why patch cannot be changed

**Pull request merged**